File: src/main/java/org/cerion/stocklist/functions/FunctionBase.java
package org.cerion.stocklist.functions;

import org.cerion.stocklist.functions.types.IFunctionEnum;

import java.util.Arrays;

public abstract class FunctionBase implements IFunction {

    private Number[] mParams;
    private final IFunctionEnum mId;

    protected FunctionBase(IFunctionEnum id, Number ...params) {
        mId = id;
        mParams = removeDoubles(params);
    }

    @Override
    public int hashCode() {
        final int prime = 31;
        int result = 1;
        result = prime * result + mId.getOrdinal();
        result = prime * result + Arrays.hashCode(mParams);
        return result;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj)
            return true;
        if (obj == null)
            return false;
        if (getClass() != obj.getClass())
            return false;

        FunctionBase other = (FunctionBase) obj;
        if (mId != other.mId)
            return false;

        // Finally, equal if parameters are equal
        return  Arrays.equals(mParams, other.mParams);
    }

    @Override
    public String toString() {
        // String.join isn't working on android with a Set, possibly java v7/8 issue
        String join = getId().toString();

        int i = 0;
        for(Number n : mParams) {
            if(i > 0)
                join += ",";
            else
                join += " ";
            join += n;
            i++;
        }

        return join;
    }

    @Override
    public Number[] params() {
        return mParams;
    }

    @Override
    public IFunctionEnum getId() {
        return mId;
    }

    protected float getFloat(int pos) {
        return mParams[pos].floatValue();
    }

    protected int getInt(int pos) {
        return mParams[pos].intValue();
    }

    public void setParams(Number[] params) {
        params = removeDoubles(params);

        if(mParams.length != params.length)
            throw new IllegalArgumentException("invalid parameter count");

        for(int i = 0; i < params.length; i++) {
            if(params[0].getClass() != mParams[0].getClass())
                throw new IllegalArgumentException("invalid parameter type at position " + i);
        }
        mParams = params;
    }

    private Number[] removeDoubles(Number[] params) {
        for(int i = 0; i < params.length; i++) {
            if(params[i].getClass() == Double.class)
                params[i] = params[i].floatValue();
        }

        return params;
    }
}

File: src/main/java/org/cerion/stocklist/functions/FunctionBase.kt
package org.cerion.stocklist.functions

import org.cerion.stocklist.arrays.ValueArray
import org.cerion.stocklist.functions.types.IFunctionEnum
import java.util.*

abstract class FunctionBase protected constructor(override val id: IFunctionEnum, vararg params: Number) : IFunction {

    private val _params: MutableList<Number>

    init {
        _params = removeDoubles(*params)
    }

    override val params: List<Number> = _params

    override fun hashCode(): Int {
        val prime = 31
        var result = 1
        result = prime * result + id.ordinal
        result = prime * result + _params.hashCode()
        return result
    }

    override fun equals(other: Any?): Boolean {
        if (this === other)
            return true
        if (other == null)
            return false
        if (other !is FunctionBase)
            return false
        if (javaClass != other.javaClass)
            return false

        return if (id !== other.id) false else _params == other._params

        // Finally, equal if parameters are equal
    }

    override fun toString(): String {
        // String.join isn't working on android with a Set, possibly java v7/8 issue
        var join = id.toString()

        var i = 0
        for (n in _params) {
            if (i > 0)
                join += ","
            else
                join += " "
            join += n
            i++
        }

        return join
    }

    protected fun getFloat(pos: Int): Float {
        return _params[pos].toFloat()
    }

    protected fun getInt(pos: Int): Int {
        return _params[pos].toInt()
    }

    override fun setParams(vararg params: Number) {
        if (_params.size != params.size)
            throw IllegalArgumentException("invalid parameter count")

        val newParams = removeDoubles(*params)

        for (i in newParams.indices) {
            val c1 = newParams[i].javaClass
            val c2 = _params[i].javaClass
            if (c1 != c2)
                throw IllegalArgumentException("invalid parameter type at position $i")
        }

        Collections.copy(_params, newParams)
    }

    override val resultType: Class<out ValueArray>
        get() {
            try {
                val methods = javaClass.methods.filter { it.name == "eval" && it.returnType.name != ValueArray::class.java.name }

                return methods.first().returnType as Class<out ValueArray>
                //val evalMethod = javaClass.getMethod("eval", FloatArray::class.java)
                //return evalMethod.returnType as Class<out ValueArray>
            } catch (e: Exception) {
                throw RuntimeException(e.message)
            }

        }

    private fun removeDoubles(vararg params: Number): MutableList<Number> {
        val result = params.toMutableList()

        for (i in result.indices) {
            if (result[i].javaClass == Double::class.java)
                result[i] = params[i].toFloat()
        }

        return result
    }
}
